### PR TITLE
[MWPW-131197] Gnav active element detection

### PR DESCRIPTION
**Pull request by overmyheadandbody**
## Description
To reach better feature parity between the AEM and Milo versions of the Global Navigation, this PR implements active element detection for navigation elements. A detailed Wiki is available in the JIRA issue, describing all use-cases.

A few developer notes, in order of files:
- test URLs for the Homepage have been added to the _gnav.md_ custom PR template;
- the `column-break` synthetic block added for https://github.com/adobecom/milo/pull/1532 has been added to _fallback.js_ to avoid false alerts when previewing navigation files;
- link groups that don't contain links will be transformed into empty strings to avoid `null` elements in the final markup;
- you might notice that on top of the `.feds-navItem--active` class, another `.feds-navItem--activeDeferred` is used. The latter is used for when no items loaded synchronously contain an active element and the navigation contains multiple asynchronously-loaded section menus. Since we depend on async code to mark the active element, we want to adapt the behaviour to avoid potential CLS issues, thus we use this extra modifier. It gets removed once the viewport switches to mobile or the content of the navigation is overflowing, since CLS should no longer be impacted.

## Related Issue
Resolves: [MWPW-131197](https://jira.corp.adobe.com/browse/MWPW-131197)

## Testing instructions
Go on a page whose URL is referenced by a link from the Global Navigation. The top navigation item that contains that link should have its text bolded and receive a bottom border.

Feel free to edit some experiences and add multiple items with the potential of being active. There should obviously be just one active item in the navigation. If multiple links match the criteria, the Wiki describes which gets priority.

## Screenshots:
Screenshot 2023-12-01 at 15 04 58

## Test URLs
**Acrobat** (link in dropdown):
- Before: https://www.stage.adobe.com/acrobat/online/sign-pdf.html?martech=off
- After: https://www.stage.adobe.com/acrobat/online/sign-pdf.html?martech=off&milolibs=gnav-active-elem--milo--overmyheadandbody

**BACOM** (simple top navigation link):
- Before: https://main--bacom--adobecom.hlx.page/products/real-time-customer-data-platform/rtcdp?martech=off
- After: https://main--bacom--adobecom.hlx.page/products/real-time-customer-data-platform/rtcdp?martech=off&milolibs=gnav-active-elem--milo--overmyheadandbody

**CC** (cloud menu):
- Before: https://main--cc--adobecom.hlx.live/?martech=off
- After: https://main--cc--adobecom.hlx.live/?martech=off&milolibs=gnav-active-elem--milo--overmyheadandbody

**Homepage** (no item):
- Before: https://main--homepage--adobecom.hlx.page/homepage/index-loggedout?martech=off
- After: https://main--homepage--adobecom.hlx.page/homepage/index-loggedout?martech=off&milolibs=gnav-active-elem--milo--overmyheadandbody

**Milo** (multiple cloud menus):
- Before: https://main--milo--overmyheadandbody.hlx.page/drafts/ramuntea/gnav-refactor-colourful?foo=bar#baz&martech=off
- After: https://gnav-active-elem--milo--overmyheadandbody.hlx.page/drafts/ramuntea/gnav-refactor-colourful?martech=off

### CODE REVIEW

**Review by mokimo — APPROVED** [review]
Cool 👍

File: libs/blocks/global-navigation/global-navigation.js [hunk 1]
```diff
@@ -531,6 +535,15 @@ class Gnav {
       this.elements.mainNav.appendChild(this.decorateMainNavItem(item, index));
     }
 
+    if (!hasActiveLink()) {
+      const sections = this.elements.mainNav.querySelectorAll('.feds-navItem--section');
+
+      if (sections.length === 1) {
```
[the remark this answers is not in the document]
Comment (overmyheadandbody) [reply libs/blocks/global-navigation/global-navigation.js:541]: If a match is not found in any of the navigation links and just a single cloud is used, then it's a safe assumption that the cloud is _sort of_ active. This is also mentioned in the Wiki. Even if it's not, we're still giving our users some sort of context as to what cloud they're navigating through. If we have multiple clouds, then we really need to check and there is a risk of not having any active element.

[the remark this answers is not in the document]
Comment (mokimo) [reply libs/blocks/global-navigation/global-navigation.js:541]: Ah yes, that's this use-case - yes, we've talked about it a couple of times, thanks for clarifying! 👍

**Review by robert-bogos — APPROVED** [review]

File: libs/blocks/global-navigation/base.css [hunk 2]
```diff
@@ -115,14 +115,51 @@
     padding: 0 12px;
   }
 
+  .feds-navItem--active > .feds-navLink {
+    position: relative;
+    font-weight: 700;
```
Comment (narcis-radu) [inline libs/blocks/global-navigation/base.css:120]: In the AEM implementation we're highlighting the active element with **bold**, but I see this does not get applied in Milo.

AEM:
Screenshot 2023-12-06 at 14 10 32

Milo:
Screenshot 2023-12-06 at 14 11 38

File: libs/blocks/global-navigation/global-navigation.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (narcis-radu) [reply libs/blocks/global-navigation/global-navigation.js:543]: @SilviuLCF - please make sure this use-case is properly tested, we'll need to have multiple navigations to test around.

File: libs/blocks/global-navigation/utilities/menu/menu.js [hunk 3]
```diff
@@ -311,6 +308,26 @@ const decorateMenu = (config) => logErrorFor(async () => {
...
+    if (getActiveLink(menuTemplate) instanceof HTMLElement) {
+      // Special handling on desktop, as content is loaded async;
+      // bolding the item text would normally push the content
+      // to the right, potentially causing CLS
+      const resetActiveState = () => {
+        config.template.style.width = '';
+        config.template.classList.remove(selectors.deferredActiveNavItem.slice(1));
+      };
+
+      if (isDesktop.matches) {
```
Comment (narcis-radu) [inline libs/blocks/global-navigation/utilities/menu/menu.js:321]: Noticed your message 
> It gets removed once the viewport switches to mobile or the content of the navigation is overflowing

but I think we should keep the bold item in mobile to be consistent with existing implementation.

File: libs/blocks/global-navigation/utilities/utilities.js [hunk 4]
```diff
@@ -162,6 +166,30 @@ export function setActiveDropdown(elem) {
...
+      if (hasActiveLink() || !(area instanceof HTMLElement)) return null;
+      const { origin, pathname } = window.location;
+      let activeLink;
+
+      [`${origin}${pathname}`, pathname].forEach((path) => {
+        if (activeLink) return;
+        activeLink = area.querySelector(`a[href = '${path}'], a[href ^= '${path}?'], a[href ^= '${path}#']`);
+      });
+
+      if (!activeLink) return null;
```
Comment (narcis-radu) [inline libs/blocks/global-navigation/utilities/utilities.js:185]: I don't think there's a risk here but still asking - would we need to set `activeLinkFound` as false for this scenario?

File: libs/blocks/global-navigation/utilities/utilities.js [hunk 5]
```diff
@@ -162,6 +166,30 @@ export function setActiveDropdown(elem) {
...
+    () => activeLinkFound,
+    (val) => { activeLinkFound = !!val; },
+    (area) => {
+      if (hasActiveLink() || !(area instanceof HTMLElement)) return null;
+      const { origin, pathname } = window.location;
+      let activeLink;
+
+      [`${origin}${pathname}`, pathname].forEach((path) => {
+        if (activeLink) return;
+        activeLink = area.querySelector(`a[href = '${path}'], a[href ^= '${path}?'], a[href ^= '${path}#']`);
```
Comment (narcis-radu) [inline libs/blocks/global-navigation/utilities/utilities.js:182]: cool 😬

File: libs/blocks/global-navigation/base.css (hunk 2 above)
Comment (narcis-radu) [reply libs/blocks/global-navigation/base.css:120]: I'm fine with keeping things as is as long as everything has been discussed. Maybe we should add a note in the JIRA ticket for future reference.

**Review by SilviuLCF — APPROVED** [review]
Testing details https://jira.corp.adobe.com/browse/MWPW-131197

**Pull request merged**